src/InstrumentationEngine/ProfilerManager.cpp

```diff
@@ -506,7 +507,9 @@ HRESULT MicrosoftInstrumentationEngine::CProfilerManager::AddInstrumentationMeth
             // InstrumentationMethod. Since we are using global lock - InstrumentationMethod can only set flags associated
             // with it in this thread.
             CInitializeInstrumentationMethodHolder initHolder(this);
-            hr = pInstrumentationMethod->Initialize(this, m_bValidateCodeSignature);
+            CComPtr<CProfilerManagerWrapper> pProfilerManagerWrapper;
+            pProfilerManagerWrapper.Attach(new (nothrow) CProfilerManagerWrapper(pInstrumentationMethod->GetClassId(), this));
+            hr = pInstrumentationMethod->Initialize(pProfilerManagerWrapper.Detach(), m_bValidateCodeSignature);
```

Comment:
I think you should just pass in pProfilerManagerWrapper without detaching.

IIRC, constructing the CProfilerManagerWrapper has the reference count set to 1. CComPtr::Attach/Detach don't change the reference count. The instrumentation method is going to AddRef and Release as needed, but nothing else is holding onto the CProfilerManagerWrapper, so it gets leaked because the initial count never gets released.

Reply:
Good catch

---
In reply to: [331294863](https://github.com/microsoft/CLRInstrumentationEngine/pull/146#discussion_r331294863) [](ancestors = 331294863)

Reply:
The current implementation with Detach fails RuntimeExceptionCallbacksTest with error -1073741819.

Removing the detach causes the RuntimeExceptionCallbacks_Debug_x64.exe to crash due to access violation during release. 

I'm going to reactivate this comment and investigate further.

---
In reply to: [331680571](https://github.com/microsoft/CLRInstrumentationEngine/pull/146#discussion_r331680571) [](ancestors = 331680571,331294863)

Reply:
Turns out the test instrumentation method doesn't addref the profilermanager so it gets released causing the issue.

---
In reply to: [331696295](https://github.com/microsoft/CLRInstrumentationEngine/pull/146#discussion_r331696295) [](ancestors = 331696295,331680571,331294863)